2016-11-17-mvcc-in-tikv.md

```diff
@@ -70,55 +70,57 @@ impl Storage {
+- **Prewrite**
 
-# TiKV-Ctl for MVCC
+  1. The transaction starts. The client obtains the current timestamp (`startTS`) from TSO.
+  2. Select one row as the primary row, the others as the secondary rows.
+  3. Check [whether there is other locks on this row](https://github.com/pingcap/tikv/blob/master/src/storage/mvcc/txn.rs#L71) or whether there are some commits located after `startTS`. These two situations will lead to conflicts. If either happens, the commit fails and [rollback](https://github.com/pingcap/tikv/blob/master/src/storage/mvcc/txn.rs#L115) will be called.
+  4. [Lock the primary row](https://github.com/pingcap/tikv/blob/master/src/storage/mvcc/txn.rs#L80).
+  5. Repeat the steps above on secondary rows.
 
-During developing and debugging, sometimes we need to know the MVCC version information.So we develop a new tool for searching the MVCC information. `TiKV` stores the Key-Values, Locks and Writes information in `CF_DEFAULT`, `CF_LOCK`, `CF_WRITE`.
-All the  values of the CF are encoded as following:
+- **Commit**
 
-|  | default | lock | write |
-| --- | --- | --- | --- |
-| **key** | z{encoded_key}{start_ts(desc)} | z{encoded_key} | z{encoded_key}{commit_ts(desc)} |
-| **value** | {value} | {flag}{primary_key}{start_ts(varint)} | {flag}{start_ts(varint)} |
+  1. Obtain the commit timestamp `commit_ts` from TSO.
+  2. Check whether the lock on the primary row still exists. Proceed if the lock is still there. Roll back if not.
+  3. Write to column `CF_WRITE` with `commit_ts`.
+  4. Clear the corresponding primary lock.
+  5. Repeat the steps above on secondary rows.
 
-Details can be found [here](https://github.com/pingcap/tikv/issues/1077).
+#### Garbage collector
 
-Since all the MVCC version information is stored as CF Key-Values in RocksDB, to search for a Key's version information, we just need to encode the key with different formats then search in the corresponding CF. The CF Key-Values are modeled by [MvccKv](https://github.com/pingcap/tikv/blob/master/src/bin/tikv-ctl.rs#L210).
+It is easy to predict that there will be more and more MVCC versions if there is no [Garbage Collector](https://github.com/pingcap/tikv/blob/master/src/storage/mvcc/txn.rs#L143) to remove the invalid versions. But we cannot simply remove all the versions before a safe point, for there may be only one version for a key, which must be kept. In TiKV, if there is any `Put` or `Delete` records before the safe point, then all the latter writes can be deleted, otherwise only `Delete`, `Rollback` and `Lock` will be deleted.
```

Comment:
```suggestion
It is easy to predict that there will be more and more MVCC versions if there is no [Garbage Collector](https://github.com/pingcap/tikv/blob/master/src/storage/mvcc/txn.rs#L143) to remove the invalid versions. But we cannot simply remove all the versions before a safe point, for there may be only one version for a key, which must be kept. In TiKV, if there is any `Put` or `Delete` records before the safe point, then all the latter writes can be deleted; otherwise only `Delete`, `Rollback` and `Lock` will be deleted.
```